src/gtk/toga_gtk/widgets/detailedlist.py

```diff
@@ -1,92 +1,185 @@
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self._on_right_click)
+
+    @property
+    def on_delete(self):
+        if self._on_delete_handler is not None:
+            return self.gtk_on_delete_clicked
+        return None
+
+    def row_factory(self, item: 'Row'):
+        new_item = TextIconRow(self.interface.factory, self, item)
+        setattr(item, '_impl', new_item)
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self._on_select_signal_handler)
+        super().destroy()
+
+    def change_source(self, source: 'ListSource'):
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We have to wait until the rows are actually added decide how to position the
+        # refresh button
+        GLib.idle_add(lambda: not self._changed())
+
+    def insert(self, index: int, item: 'Row'):
+        self.store.insert(index, self.row_factory(item))
+        self.list_box.show_all()
+        self._changed()
+
+    def change(self, item: 'Row'):
+        new_item = self.row_factory(item)
+        index = self.index(item)
+        self.store.insert(index, new_item)
+        
+    def remove(self, item: 'Row', index: int):
+        if index is None:
+            index = self.index(item)
+
+        if self._active_row == item._impl:
+            self._active_row = None
+
+        self.store.remove(index)
+        item._impl.destroy()
+        self._changed()
+        
+    def clear(self):
+        self.store.remove_all()
+        self._changed()
 
     def get_selection(self):
-        tree_model, tree_iter = self.selection.get_selected()
-        if tree_iter:
-            row = tree_model.get(tree_iter, 0)[0]
+        item = self.list_box.get_selected_row()
+        if item is None:
+            return item
         else:
-            row = None
+            return item.interface
+
+    def scroll_to_row(self, row: int):
+        item = self.store[row]
+        item.scroll_to_center()
 
-        return row
+    def set_on_refresh(self, handler: callable):
+        if handler is not None:
+            self._on_refresh_handler = handler
+            self.refresh_button.set_on_refresh(self.gtk_on_refresh_clicked)
 
-    def set_on_select(self, handler):
+    def set_on_select(self, handler: callable):
+        self._on_select_handler = handler
+        if handler is not None:
+            self.list_box.set_selection_mode(Gtk.SelectionMode.SINGLE)
+
+    def set_on_delete(self, handler: callable):
+        self._on_delete_handler = handler
+
+    def after_on_refresh(self):
         # No special handling required
         pass
 
-    def set_on_delete(self, handler):
-        pass
+    def index(self, item: 'Row'):
+        for index in range(0, len(self.store)):
```

Comment:
It's not clear why the *widget* is doing this iteration. The source can give you the position of the item in the list (and also gives the option of optimising the lookup).

Reply:
Neither `Gtk.ListBox` nor `Gio.ListStore` have an `index` method so I assumed it simply wasn't available. Turns out `Gtk.ListBoxRow` has `get_index`. That loop was really upsetting me.